src/authInfo.ts

```diff
@@ -515,9 +515,14 @@ export class AuthInfo extends AsyncCreatable<AuthInfo.Options> {
    * @param sfdxAuthUrl
    */
   public static parseSfdxAuthUrl(sfdxAuthUrl: string) {
-    const match = sfdxAuthUrl.match(
-      /^force:\/\/([a-zA-Z0-9._-]+):([a-zA-Z0-9._-]*):([a-zA-Z0-9._-]+)@([a-zA-Z0-9._-]+)/
-    );
+    let url;
+    try {
+      url = JSON.parse(sfdxAuthUrl).sfdxAuthUrl;
```

Comment:
This should be fixed in plugin-auth: https://github.com/salesforcecli/plugin-auth/blob/main/src/commands/auth/sfdxurl/store.ts#L55

Something like this:

```typescript
const sfdxAuthUrl = this.flags.sfdxurlfile.endsWith('.json')
  ? (await fs.readJson(this.flags.sfdxurlfile)).sfdxAuthUrl
  : await fs.readFile(this.flags.sfdxurlfile, 'utf8');
```

Reply:
@mdonnalley I did try that but it got pretty messy because of types.
You have to end up doing something like this
```
// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call
const sfdxAuthUrl = this.flags.sfdxurlfile.endsWith('.json')
  ? (await fs.readJson(this.flags.sfdxurlfile)).sfdxAuthUrl
  : await fs.readFile(this.flags.sfdxurlfile, 'utf8');
```

And even then `tsc` doesn't like it so it won't compile
![image](https://user-images.githubusercontent.com/1084688/110992243-a63d1300-8343-11eb-972d-85d33f361d1f.png)


I just gave it another shot though and this is what I've ended up with.
```
    let sfdxAuthUrl;
    // eslint-disable-next-line @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-member-access
    if (this.flags.sfdxurlfile.endsWith('json')) {
      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
      sfdxAuthUrl = await fs.readJson(this.flags.sfdxurlfile);
      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access
      sfdxAuthUrl = sfdxAuthUrl.sfdxAuthUrl;
    } else {
      sfdxAuthUrl = await fs.readFile(this.flags.sfdxurlfile, 'utf8');
    }
```    

I would rather avoid something that ugly, do you have any other suggestions on how to do this without having  to resort to that mess?

Reply:
Here's a PR with those changes for easier viewing
https://github.com/salesforcecli/plugin-auth/pull/81

Reply:
@RodEsp 

```typescript
 import { AnyJson } from '@salesforce/ts-types';

 type AuthUrlJson = AnyJson & {
  sfdxUrlJson: string;
 };

 public async run(): Promise<AuthFields> {
    const authFile = this.flags.sfdxurlfile as string;
    const sfdxAuthUrl = authFile.endsWith('.json')
      ? await this.getUrlFromJson(authFile)
      : await this.getUrlFromFile(authFile);
    // etc...  
  }


  private async getUrlFromJson(authFile: string): Promise<string> {
    const contents = (await fs.readJson(authFile)) as AuthUrlJson;
    return contents.sfdxUrlJson;
  }

  private async getUrlFromFile(authFile: string): Promise<string> {
    const contents = await fs.readFile(authFile, 'utf8');
    return contents;
  }
```

Reply:
Oh man, do we really want to create a whole 'nother type wiht private methods just for reading a simple JSON file? `sfdx-cli` is my first real foray into TS but is this just "the typescript way"?
This is what I meant [here](https://platformcloud.slack.com/archives/G4Q6L9KDH/p1615575817258600?thread_ts=1615575689.258400&cid=G4Q6L9KDH).

Thoughts? @WillieRuemmele @amphro @peternhale @mshanemc

Reply:
Since this is so simple I would use `Dictionary` instead of `AnyJson`

so 
```
const contents = (await fs.readJson(authFile)) as Disctiony;
return contents.sfdxUrlJson;
```

Reply:
I like Mike's approach, except I don't think the `getUrlFromFile` method is necessary;  Just do the `await readFile`.  And the AuthUrlJson type can just be `{ sfdxUrlJson: string };`